### PR TITLE
Changes for Nebula Pad, fix to layer progress, and minor UI fixes

### PR DESCRIPTION
**Pull request by ajs123**
Changes for proper display on the Nebula Pad, and a few small fixes for known issues:

- Adjusted screen limits for proper touchscreen coordinates (lv_drv_conf.h)
- Fixed layer progress update (print status panel)
- Reduced flow display to 2 digits after the decimal (print status panel)
- Changed the default font in lv_conf.h (Could this break other builds? Font setup in main.cpp doesn't seem to have effect.)
- Swapped icons for Z+ and Z- in fine tune and homing panels
- Updated build instructions for spdlog

### CODE REVIEW

**Comment by adec** [comment]
The modified instructions for cross-compilation were top notch - thanks @ajs123.

File: DEVELOPMENT.md [hunk 1]
```diff
@@ -45,29 +45,44 @@ To build guppyscreen for Mipsel (Ingenic X2000E) - specific to the K1 SoC, you w
...
-4. `make spdlogclean && make libspdlog.a`
-5. `make clean && make -j$(nproc)`
+Environment and toolchain
+
+```
+export PATH=<path-to-mips-toolchain>:$PATH
+export CROSS_COMPILE=mips-linux-gnu-
+export CC=mips-linux-gnu-gcc
+export CXX=mips-linux-gnu-g++
+export GUPPY_SMALL_SCREEN=1
```
Comment (ballaswag) [inline DEVELOPMENT.md:62]: Explanation of exporting GUPPY_SMALL_SCREEN/EVDEV_CALIBRATE will be a good information.

File: DEVELOPMENT.md [hunk 2]
```diff
@@ -45,29 +45,44 @@ To build guppyscreen for Mipsel (Ingenic X2000E) - specific to the K1 SoC, you w
...
+export EVDEV_CALIBRATE=1
+```
+
+Build libraries and executable
+
+1. `make wpaclean && make wpaclient`
```
Comment (ballaswag) [inline DEVELOPMENT.md:68]: @pellcorp actually simplified this instruction. The steps are already in the Makefile `build` target. See here https://github.com/ballaswag/guppyscreen/pull/45/files#diff-fc2a310fcfedfefb0046def697f932def80cbb1e78c689bc0af3c8eab3e33eceR56

File: lv_conf.h [hunk 3]
```diff
@@ -376,7 +376,7 @@ extern uint32_t custom_tick_get(void);
 #define LV_FONT_CUSTOM_DECLARE
 
 /*Always set a default font*/
-#define LV_FONT_DEFAULT &lv_font_montserrat_16
+#define LV_FONT_DEFAULT &lv_font_montserrat_12  // Appropriate size for Nebula Pad
```
Comment (ballaswag) [inline lv_conf.h:379]: Sorry about this nonsense, the issue was a bug in lvgl which I patched locally but didn't commit. The DPI font size is fixed by https://github.com/ballaswag/guppyscreen/commit/237c61aff3c7eaedb6a3653d907f0867a7c285c3

File: src/finetune_panel.cpp [hunk 4]
```diff
@@ -22,8 +22,8 @@ FineTunePanel::FineTunePanel(KWebSocketClient &websocket_client, std::mutex &l)
   , panel_cont(lv_obj_create(lv_scr_act()))
   , values_cont(lv_obj_create(panel_cont))
   , zreset_btn(panel_cont, &refresh_img, "Reset Z", &FineTunePanel::_handle_zoffset, this)
-  , zup_btn(panel_cont, &z_closer, "Z+", &FineTunePanel::_handle_zoffset, this)
-  , zdown_btn(panel_cont, &z_farther, "Z-", &FineTunePanel::_handle_zoffset, this)
+  , zup_btn(panel_cont, &z_farther, "Z+", &FineTunePanel::_handle_zoffset, this)
```
Comment (ballaswag) [inline src/finetune_panel.cpp:25]: This should probably be an option for the user invert these 2 icons.

File: src/print_status_panel.cpp [hunk 5]
```diff
@@ -538,7 +538,7 @@ void PrintStatusPanel::update_layers(json &info) {
     new_total_layer = v.template get<int>();
   }
 
-  if (new_total_layer > total_layer || new_cur_layer > cur_layer) {
+  if (new_total_layer != total_layer || new_cur_layer != cur_layer) {
```
Comment (ballaswag) [inline src/print_status_panel.cpp:541]: nice

File: lv_drv_conf.h [hunk 6]
```diff
@@ -455,10 +455,10 @@
 #  define EVDEV_CALIBRATE         1               /*Scale and offset the touchscreen coordinates by using maximum and minimum values for each axis*/
 
 #  if EVDEV_CALIBRATE
-#    define EVDEV_HOR_MIN         0               /*to invert axis swap EVDEV_XXX_MIN by EVDEV_XXX_MAX*/
-#    define EVDEV_HOR_MAX      4096               /*"evtest" Linux tool can help to get the correct calibraion values>*/
-#    define EVDEV_VER_MIN         0
-#    define EVDEV_VER_MAX      4096
+#    define EVDEV_HOR_MIN         190 //0               /*to invert axis swap EVDEV_XXX_MIN by EVDEV_XXX_MAX*/
```
Comment (ballaswag) [inline lv_drv_conf.h:458]: Thanks for the numbers. This needs to be configurable so other screens that needs calibration do not get impacted (e.g. the BTT Pad 7).

**Comment by Rulevoid** [comment]
This is a pretty massive fix for many issues with Guppy on Nebula pads.  Looks like this has been sitting for a while. Are there some easy to follow steps for me to compile and use this branch on my Ender 3 V3 KE w/ Nebula Pad? (Already using Guppy on it)?

File: src/finetune_panel.cpp (hunk 4 above)
Comment (Rulevoid) [reply src/finetune_panel.cpp:25]: The icons are pretty industry standard if I'm not mistaken. Flipping them (as they currently are) could be potentially dangerous in terms of leading to nozzle crashes (happened to me).

**Comment by ballaswag** [comment]
Fixes in nightly build here https://github.com/ballaswag/guppyscreen/actions/runs/8629364623 guppyscreen-smallscreen. If anyone tests it out please let me know if touch calibration works.

**Comment by ajs123** [comment]
> Fixes in nightly build here https://github.com/ballaswag/guppyscreen/actions/runs/8629364623 guppyscreen-smallscreen. If anyone tests it out please let me know if touch calibration works.

The screen cal worked for me!

**Comment by Rulevoid** [comment]
@ajs123 How did you get the file to an Ender 3 V3 KE? I don't think they support FTP (WinSCP is failing for me).

**Comment by ajs123** [comment]
> @ajs123 How did you get the file to an Ender 3 V3 KE? I don't think they support FTP (WinSCP is failing for me).

On the build machine...
`scp build/bin/guppyscreen root@<ip.address>:/usr/data/guppyscreen/guppyscreen-new`
(may require the -O option to scp)

Then on the Nabula...
`cd /usr/data/guppyscreen`
`cp guppyscreen guppyscreen-save`
`cp guppydcreen-new guppyscreen`
`/etc/init.d/S99guppyscreen restart`

**Comment by Rulevoid** [comment]
Thanks! Only difference I had was to make the permissions match on the new binary vs. the previous one.

My thoughts:
This build still looks like it has some issues. The Homing screen still has the icons for the Z+ and Z- backwards, the target temperature for extruder on the main screen is still wrapping lines, etc. The good is that the refresh button under the Print icon now actually refreshes the gcode directory, and the touchscreen calibration has made more accurate clicks.

image

image

**Pull request closed without merging**